### ISSUE TITLE
Webdataset tar shuffling

### ISSUE THREAD

**Issue by ZeWang95** [awslabs/s3-connector-for-pytorch#247]
### Tell us more about this new feature.

Hi! Thanks for the great work!

I'm using s3torchconnector with webdataset. I'm initializing the dataset by:
s3_dataset = s3torchconnector.S3IterableDataset.from_prefix( SHARDS_URI, .......
where SHARDS_URI is the s3 path containing all the tar files. 

I'm wondering if there is any way I can shuffle the tar files after to make sure of the randomness. I'm not using torch.utils.data.DataLoader so I'd like to make sure that the tar files are shuffled properly. 

Thanks!
Ze

**Comment by IsaevIlya** [comment]
Hi @ZeWang95,
Thank you for your interest in using s3-connector-for-pytorch and for raising this important question about shuffling data. We appreciate your feedback, as it helps us improve our library and better serve the community.

The s3-connector-for-pytorch library itself does not currently provide built-in support for shuffling the order in which objects are read from S3. However, PyTorch offers several utilities that can be used in conjunction with the `S3IterableDataset` and `webdataset` to achieve this functionality. One approach is to use the [ShufflerIterDataPipe](https://github.com/pytorch/pytorch/blob/main/torch/utils/data/datapipes/iter/combinatorics.py#L61) from PyTorch's datapipe API. This utility maintains a buffer of samples from the input iterator and randomly shuffles them before yielding them to the downstream components. Here's an example of how you can use it: 

```python
s3_dataset = s3torchconnector.S3IterableDataset.from_prefix(SHARDS_URI...)
tar_dataset = webdataset.tariterators.tar_file_expander(s3_dataset)
# Shuffle the WebDataset using ShufflerIterDataPipe
buffer_size = 1000  # Adjust this value based on your memory constraints
shuffled_dataset = ShufflerIterDataPipe(tar_dataset, buffer_size=buffer_size)

# You can now iterate over shuffled_dataset
for sample in shuffled_dataset:
    # Process the sample
    ...
```

In this example, the `ShufflerIterDataPipe` will maintain a buffer of 1000 samples from tar_dataset and randomly shuffle them before yielding them to the downstream components. Larger buffer sizes generally lead to better shuffling at the cost of higher memory usage. You should adjust this value based on your available memory resources. This approach ensures that the tar files are read in a randomized order, providing the desired level of shuffling.

Please note that this approach shuffles the samples within the tar files, not the order in which the tar files themselves are read from S3. If you need to randomize the order of tar files, you may need to implement additional logic to retrieve a shuffled list of tar file paths from S3 and then utilize `S3IterableDataset.from_objects` instead of `S3IterableDataset.from_prefix`.

We hope this explanation helps you. If you have any further questions or need additional assistance, please don't hesitate to reach out.

Ilya

**Comment by ZeWang95** [comment]
Thank you! I think from_objects can solve the problem well!
